### PR TITLE
Enhancement/2421 query analytics reports dimension filters

### PR DESCRIPTION
**Pull request by ivankruchkoff**
## Summary

Addresses issue #2421 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/modules/analytics/util/report-validation.js [hunk 1]
```diff
@@ -57,3 +57,38 @@ export function isValidDimensions( dimensions ) {
...
+
+	if ( Array.isArray( dimensions ) ) {
+		dimensions.forEach( ( singleDimension ) => parseDimension( singleDimension, validDimensions ) );
+	} else {
+		parseDimension( dimensions, validDimensions );
+	}
+
+	// Ensure every dimensionFilter key corresponds to a valid dimension and ensure every value is a string.
+	return Object.keys( dimensionFilters ).every( ( dimensionFilterKey ) => validDimensions.include( dimensionFilterKey ) ) &&
+		Object.values( dimensionFilters ).every( ( dimensionFilterValue ) => typeof dimensionFilterValue === 'string' && dimensionFilterValue.length );
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics/util/report-validation.js:93]: @ivankruchkoff I think we don't need to verify that all values are non-empty strings because there might be a case when we want to pass numeric values for certain dimensions. So, let's change it to check only keys.

File: assets/js/modules/analytics/util/report-validation.js [hunk 2]
```diff
@@ -57,3 +57,38 @@ export function isValidDimensions( dimensions ) {
...
+	};
+
+	if ( Array.isArray( dimensions ) ) {
+		dimensions.forEach( ( singleDimension ) => parseDimension( singleDimension, validDimensions ) );
+	} else {
+		parseDimension( dimensions, validDimensions );
+	}
+
+	// Ensure every dimensionFilter key corresponds to a valid dimension and ensure every value is a string.
+	return Object.keys( dimensionFilters ).every( ( dimensionFilterKey ) => validDimensions.include( dimensionFilterKey ) ) &&
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics/util/report-validation.js:92]: I believe we can simplify it to this:

```suggestion
	return Object.keys( dimensionFilters ).every( validDimensions.include, validDimensions ) &&
```

File: includes/Modules/Analytics.php [hunk 3]
```diff
@@ -962,6 +962,33 @@ function ( $dimension_def ) {
...
 				}
 
+				$dimension_filters          = $data['dimensionFilters'];
+				$dimension_filter_instances = array();
+				if ( ! empty( $dimension_filters ) && ( is_string( $dimension_filters ) || is_array( $dimension_filters ) ) ) {
+					if ( is_string( $dimension_filters ) ) {
+						$dimension_filters = explode( ',', $dimension_filters );
+					} elseif ( is_array( $dimension_filters ) && ! wp_is_numeric_array( $dimension_filters ) ) { // If single object is passed.
+						$dimension_filters = array( $dimension_filters );
+					}
```
Comment (eugene-manuilov) [inline includes/Modules/Analytics.php:972]: @ivankruchkoff `$data['dimensionFilters']` should always be an associated array (see your `isValidDimensionFilters` function), so we don't need this checks. If the `dimensionFilters` is not an array, we can ignore it.

File: includes/Modules/Analytics.php [hunk 4]
```diff
@@ -1379,14 +1418,19 @@ protected function create_analytics_site_data_request( array $args = array() ) {
 			$request->setDateRanges( array( $date_range ) );
 		}
 
+		if ( ! empty( $args['dimension_filters'] ) ) {
+			$request->setDimensions( (array) $args['dimension_filters'] );
+		}
+
```
Comment (eugene-manuilov) [inline includes/Modules/Analytics.php:1424]: @ivankruchkoff do we really need to set dimensions here using `$args['dimension_filters']`? We already set dimensions using `$args['dimensions']` a few steps above.

File: includes/Modules/Analytics.php [hunk 5]
```diff
@@ -1379,14 +1418,19 @@ protected function create_analytics_site_data_request( array $args = array() ) {
...
 			$dimension_filter->setDimensionName( 'ga:pagePath' );
 			$dimension_filter->setOperator( 'EXACT' );
 			$args['page'] = str_replace( trim( $this->context->get_reference_site_url(), '/' ), '', esc_url_raw( $args['page'] ) );
 			$dimension_filter->setExpressions( array( rawurldecode( $args['page'] ) ) );
-			$dimension_filter_clause = new Google_Service_AnalyticsReporting_DimensionFilterClause();
-			$dimension_filter_clause->setFilters( array( $dimension_filter ) );
+			$dimension_filters_clauses[] = $dimension_filter;
+			$dimension_filter_clause     = new Google_Service_AnalyticsReporting_DimensionFilterClause();
+			$dimension_filter_clause->setFilters( $dimension_filters_clauses );
 			$request->setDimensionFilterClauses( array( $dimension_filter_clause ) );
```
Comment (eugene-manuilov) [inline includes/Modules/Analytics.php:1434]: @ivankruchkoff we need to move this out of the `if ( ! empty( $args['page'] ) )` condition to make sure that dimension filters work even if the page is not empty.

File: assets/js/modules/analytics/datastore/report.js [hunk 6]
```diff
@@ -72,6 +72,13 @@ const fetchGetReportStore = createFetchStore( {
 			);
 		}
 
+		if ( dimensionFilters ) {
+			invariant(
+				isValidDimensionFilters( dimensionFilters, dimensions ),
+				'Dimensions filters must be an object where the keys are valid Dimensions and values are non-zero length strings.',
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics/datastore/report.js:78]: We probably need to adjust this message since we expect values not to be just non-zero strings.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @ivankruchkoff! I have found a few things that we need to address before I can approve this PR. Could you please take a look at it?

File: includes/Modules/Analytics.php (hunk 4 above)
Comment (ivankruchkoff) [reply includes/Modules/Analytics.php:1424]: Removed.

File: assets/js/modules/analytics/util/report-validation.js (hunk 1 above)
Comment (ivankruchkoff) [reply assets/js/modules/analytics/util/report-validation.js]: Addressed

File: includes/Modules/Analytics.php (hunk 3 above)
Comment (ivankruchkoff) [reply includes/Modules/Analytics.php]: Addressed.

File: assets/js/modules/analytics/datastore/report.js (hunk 6 above)
Comment (ivankruchkoff) [reply assets/js/modules/analytics/datastore/report.js]: Addressed.

File: includes/Modules/Analytics.php (hunk 5 above)
Comment (ivankruchkoff) [reply includes/Modules/Analytics.php:1424]: Perhaps I was reading too literally into AC ```This dimension filter clause then needs to be combined with the existing dimension filter clause based on the page query parameter (see Analytics::create_analytics_site_data_request).```

Comment (ivankruchkoff) [reply includes/Modules/Analytics.php]: Addressed.

**Review by eugene-manuilov — DISMISSED** [review]
Awesome! Thanks, @ivankruchkoff!

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @ivankruchkoff – this is looking good overall, just a few changes to address.

File: assets/js/modules/analytics/datastore/report.js [hunk 7]
```diff
@@ -57,7 +57,7 @@ const fetchGetReportStore = createFetchStore( {
 		invariant( isPlainObject( options ), 'Options for Analytics report must be an object.' );
 		invariant( isValidDateRange( options ), 'Either date range or start/end dates must be provided for Analytics report.' );
 
-		const { metrics, dimensions, orderby } = normalizeReportOptions( options );
+		const { metrics, dimensions, dimensionFilters, orderby } = normalizeReportOptions( options );
```
[the remark this answers is not in the document]
Comment (ivankruchkoff) [reply assets/js/modules/analytics/datastore/report.js:60]: I've add a function to normalize dimensions in https://github.com/google/site-kit-wp/pull/2485/commits/a439154ad17f08504a06fb79068d10ff48fe7df8 however not sure what the dimensionFilters possible unnormalized values would be as it's a new param we're introducing. Any thoughts @aaemnnosttv ?

[the remark this answers is not in the document]
Comment (ivankruchkoff) [reply assets/js/modules/analytics/datastore/report.js:60]: The best I could think of is https://github.com/google/site-kit-wp/pull/2485/commits/e623d0317ebea23a001043f817233df3fbe02dd1

File: assets/js/modules/analytics/util/report-validation.js [hunk 8]
```diff
@@ -57,3 +57,37 @@ export function isValidDimensions( dimensions ) {
...
+/**
+ * Verifies provided dimensionFilters to make sure they match allowed values found in dimensions.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Object}                               dimensionFilters The dimension filters to check.
+ * @param {string|string[]|Object|Object[]|null} dimensions       The dimensions to check.
+ * @return {boolean} TRUE if dimension filters are valid, otherwise FALSE.
+ */
+export function isValidDimensionFilters( dimensionFilters, dimensions ) {
```
[the remark this answers is not in the document]
Comment (ivankruchkoff) [reply assets/js/modules/analytics/util/report-validation.js:75]: Added in https://github.com/google/site-kit-wp/pull/2485/commits/df1676361aa88f1cbff68dce9e8b72a783a15bed

File: assets/js/modules/analytics/datastore/report.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/analytics/datastore/report.js:60]: Re: normalizing `dimensionFilters`, I was mostly referring to the type since I'm not exactly sure what the possible values can be here, but if the types are limited then I think what you have is good 👍

File: includes/Modules/Analytics.php [hunk 9]
```diff
@@ -962,6 +962,27 @@ function ( $dimension_def ) {
 					}
 				}
 
+				$dimension_filters          = $data['dimensionFilters'];
+				$dimension_filter_instances = array();
+				if ( ! empty( $dimension_filters ) && is_array( $dimension_filters ) ) {
+					foreach ( $request_args['dimensions'] as $dimension ) {
+						$dimension_name = $dimension->getName();
+						if ( isset( $dimension_filters[ $dimension_name ] ) ) {
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics.php:970]: Let's avoid the additional nesting here by using `continue` if the key isn't set.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @ivankruchkoff – this is very close, just a few final things to touch up then this should be good to go.

File: assets/js/modules/analytics/util/report-normalization.js [hunk 10]
```diff
@@ -51,3 +53,26 @@ const normalizeMetrics = ( metrics ) => {
...
+
+const normalizeDimensionFilters = ( dimensionFilters ) => {
+	const normalizedFilters = {};
+	if ( isPlainObject( dimensionFilters ) ) {
+		for ( const [ key, value ] of Object.entries( dimensionFilters ) ) {
+			if ( [ 'string', 'number', 'boolean' ].includes( typeof value ) ) {
+				normalizedFilters[ key ] = value;
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (ivankruchkoff) [reply assets/js/modules/analytics/util/report-normalization.js:76]: Updated.

File: includes/Modules/Analytics.php (hunk 9 above)
Comment (ivankruchkoff) [reply includes/Modules/Analytics.php]: Updated.

File: assets/js/modules/analytics/util/report-validation.test.js [hunk 11]
```diff
@@ -64,6 +64,35 @@ describe( 'Analytics Reporting API validation', () => {
 		} );
 	} );
 
+	describe( 'isValidDimensionFilters', () => {
+		it( 'should return TRUE if a valid object is passed with a valid dimension', () => {
+			expect( isValidDimensionFilters( {
+				test: 'foo',
+			}, [ {
+				name: 'test',
+			} ] ) ).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (ivankruchkoff) [reply assets/js/modules/analytics/util/report-validation.test.js]: Updated.

File: includes/Modules/Analytics.php [hunk 12]
```diff
@@ -1335,27 +1358,30 @@ function ( Google_Service_Analytics_Account $account ) {
...
 	 * @param array $args {
 	 *     Optional. Additional arguments.
 	 *
-	 *     @type array  $dimensions List of request dimensions. Default empty array.
-	 *     @type string $start_date Start date in 'Y-m-d' format. Default empty string.
-	 *     @type string $end_date   End date in 'Y-m-d' format. Default empty string.
-	 *     @type string $page       Specific page URL to filter by. Default empty string.
-	 *     @type int    $row_limit  Limit of rows to return. Default 100.
+	 *     @type array  $dimensions        List of request dimensions. Default empty array.
+	 *     @type array  $dimension_filters List of dimension filters for the specified request dimensions. Default empty array.
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics.php:1367]: Let's add the type here as it isn't quite clear that this argument requires `Google_Service_AnalyticsReporting_DimensionFilter` instances.

```suggestion
	 *     @type Google_Service_AnalyticsReporting_DimensionFilter[]  $dimension_filters List of dimension filter instances for the specified request dimensions. Default empty array.
```

File: includes/Modules/Analytics.php [hunk 13]
```diff
@@ -962,6 +962,29 @@ function ( $dimension_def ) {
...
+						if ( ! isset( $dimension_filters[ $dimension_name ] ) ) {
+							continue;
+						}
+
+						$dimension_value  = $dimension_filters[ $dimension_name ];
+						$dimension_filter = new Google_Service_AnalyticsReporting_DimensionFilter();
+						$dimension_filter->setDimensionName( $dimension_name );
+						$dimension_filter->setOperator( 'EXACT' );
+						$dimension_filter->setExpressions( array( $dimension_value ) );
+						$dimension_filter->setCaseSensitive( false );
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics.php:979]: Let's remove this since the default is `false` and it isn't mentioned in the AC.
```suggestion
```

File: includes/Modules/Analytics.php [hunk 14]
```diff
@@ -1385,8 +1420,12 @@ protected function create_analytics_site_data_request( array $args = array() ) {
 			$dimension_filter->setOperator( 'EXACT' );
 			$args['page'] = str_replace( trim( $this->context->get_reference_site_url(), '/' ), '', esc_url_raw( $args['page'] ) );
 			$dimension_filter->setExpressions( array( rawurldecode( $args['page'] ) ) );
+			$dimension_filters_clauses[] = $dimension_filter;
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics.php:1423]: This is incorrect – only the `Google_Service_AnalyticsReporting_DimensionFilterClause` should be added to this list, not the `Google_Service_AnalyticsReporting_DimensionFilter` instance (as it is in the current version of the file on the base branch).

File: includes/Modules/Analytics.php [hunk 15]
```diff
@@ -1385,8 +1420,12 @@ protected function create_analytics_site_data_request( array $args = array() ) {
...
+		}
+
+		if ( ! empty( $dimension_filters_clauses ) ) {
 			$dimension_filter_clause = new Google_Service_AnalyticsReporting_DimensionFilterClause();
-			$dimension_filter_clause->setFilters( array( $dimension_filter ) );
+			$dimension_filter_clause->setFilters( $dimension_filters_clauses );
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics.php:1428]: This clause should remain independent as it is in develop now, but `setFilters` should be called with `array( $dimension_filter )` like it was before rather than the array of clauses.

File: includes/Modules/Analytics.php [hunk 16]
```diff
@@ -1385,8 +1420,12 @@ protected function create_analytics_site_data_request( array $args = array() ) {
...
 			$request->setDimensionFilterClauses( array( $dimension_filter_clause ) );
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics.php:1429]: This is where the array of clauses should be set 😄 , although this condition should remain separate from the conditions that define the clauses themselves.
```suggestion
			$request->setDimensionFilterClauses( $dimension_filters_clauses );
```

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @ivankruchkoff ! This is pretty close, just a few details to square away on the PHP side 👍

File: includes/Modules/Analytics.php (hunk 14 above)
Comment (ivankruchkoff) [reply includes/Modules/Analytics.php:1422]: Yes, you're correct.

File: includes/Modules/Analytics.php (hunk 15 above)
Comment (ivankruchkoff) [reply includes/Modules/Analytics.php:1427]: That's right, changed it back.

File: includes/Modules/Analytics.php [hunk 17]
```diff
@@ -1335,27 +1357,30 @@ function ( Google_Service_Analytics_Account $account ) {
...
-	 *     @type string $start_date Start date in 'Y-m-d' format. Default empty string.
-	 *     @type string $end_date   End date in 'Y-m-d' format. Default empty string.
-	 *     @type string $page       Specific page URL to filter by. Default empty string.
-	 *     @type int    $row_limit  Limit of rows to return. Default 100.
+	 *     @type array  $dimensions        List of request dimensions. Default empty array.
+	 *     @type Google_Service_AnalyticsReporting_DimensionFilter[]  $dimension_filters List of dimension filter instances for the specified request dimensions. Default empty array.
+	 *     @type string $start_date        Start date in 'Y-m-d' format. Default empty string.
+	 *     @type string $end_date          End date in 'Y-m-d' format. Default empty string.
+	 *     @type string $page              Specific page URL to filter by. Default empty string.
+	 *     @type int    $row_limit         Limit of rows to return. Default 100.
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics.php:1370]: These need to be aligned now after the `@type` change for `$dimension_filters`. Odd that PHPCS doesn't flag it - it must be this "inner" declaration isn't a linted the same way. No worries.

```suggestion
	 *     @type array                                               $dimensions        List of request dimensions. Default empty array.
	 *     @type Google_Service_AnalyticsReporting_DimensionFilter[] $dimension_filters List of dimension filter instances for the specified request dimensions. Default empty array.
	 *     @type string                                              $start_date        Start date in 'Y-m-d' format. Default empty string.
	 *     @type string                                              $end_date          End date in 'Y-m-d' format. Default empty string.
	 *     @type string                                              $page              Specific page URL to filter by. Default empty string.
	 *     @type int                                                 $row_limit         Limit of rows to return. Default 100.
```

File: includes/Modules/Analytics.php [hunk 18]
```diff
@@ -1385,9 +1419,12 @@ protected function create_analytics_site_data_request( array $args = array() ) {
 			$dimension_filter->setOperator( 'EXACT' );
 			$args['page'] = str_replace( trim( $this->context->get_reference_site_url(), '/' ), '', esc_url_raw( $args['page'] ) );
 			$dimension_filter->setExpressions( array( rawurldecode( $args['page'] ) ) );
+		}
+
+		if ( ! empty( $dimension_filters_clauses ) ) {
 			$dimension_filter_clause = new Google_Service_AnalyticsReporting_DimensionFilterClause();
 			$dimension_filter_clause->setFilters( array( $dimension_filter ) );
-			$request->setDimensionFilterClauses( array( $dimension_filter_clause ) );
+			$request->setDimensionFilterClauses( $dimension_filters_clauses );
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics.php:1427]: This is still not quite right (no better place to leave this comment although this specific line _is_ correct).

The `$dimension_filter_clause` immediately above belongs in the preceding conditional where the filter is defined.

This conditional should only have the call to `$request->setDimensionFilterClauses()`.

**Review by aaemnnosttv — COMMENTED** [review]
Almost there, just a few small but important details left to fix. I'll push a few commits to address these quickly.

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

**Review by aaemnnosttv — APPROVED** [review]
Super!

File: includes/Modules/Analytics.php [hunk 19]
```diff
@@ -1368,6 +1393,15 @@ protected function create_analytics_site_data_request( array $args = array() ) {
 		$request->setIncludeEmptyRows( true );
 		$request->setViewId( $profile_id );
 
+		$dimension_filter_clauses = array();
+		if ( ! empty( $args['dimension_filters'] ) ) {
+			$dimension_filters       = $args['dimension_filters'];
+			$dimension_filter_clause = new Google_Service_AnalyticsReporting_DimensionFilterClause();
+			$dimension_filter_clause->setFilters( array( $dimension_filters ) );
```
Comment (felixarntz) [inline includes/Modules/Analytics.php:1400]: I believe this is incorrect? `$dimension_filters` should already be passed as the array required here, so it must not be wrapped again.

File: includes/Modules/Analytics.php [hunk 20]
```diff
@@ -1368,6 +1393,15 @@ protected function create_analytics_site_data_request( array $args = array() ) {
...
+			$dimension_filter_clause->setOperator( 'AND' );
+			$dimension_filter_clauses[] = $dimension_filter_clause;
+		}
+
```
Comment (felixarntz) [inline includes/Modules/Analytics.php:1404]: Nitpick, but maybe move this a bit further down to be right above the `if ( ! empty( $args['page'] ) ) {` clause? This way all logic touching dimension filter clauses is located close together.

File: includes/Modules/Analytics.php [hunk 21]
```diff
@@ -962,6 +962,28 @@ function ( $dimension_def ) {
 					}
 				}
 
+				$dimension_filters          = $data['dimensionFilters'];
+				$dimension_filter_instances = array();
+				if ( ! empty( $dimension_filters ) && is_array( $dimension_filters ) ) {
+					foreach ( $request_args['dimensions'] as $dimension ) {
```
Comment (felixarntz) [inline includes/Modules/Analytics.php:968]: This is not correct. `dimensionFilters` and `dimensions` should be completely decoupled, the arguments are not related, unless I'm missing something. To validate/normalize the `dimensionFilters` argument, you should only need to look at `dimensionFilters`, not `dimensions`.

* `dimensionFilters` ensures only results that satisfy a certain dimension criteria are included in the response.
* `dimensions` modifies the response format to group results by certain dimensions.

File: assets/js/modules/analytics/util/report-validation.js [hunk 22]
```diff
@@ -57,3 +57,21 @@ export function isValidDimensions( dimensions ) {
...
+/**
+ * Verifies provided dimensionFilters to make sure they match allowed values found in dimensions.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Object}   dimensionFilters The dimension filters to check.
+ * @param {Object[]} dimensions       The dimensions to check.
+ * @return {boolean} TRUE if dimension filters are valid, otherwise FALSE.
+ */
+export function isValidDimensionFilters( dimensionFilters, dimensions ) {
```
Comment (felixarntz) [inline assets/js/modules/analytics/util/report-validation.js:70]: See my other comment, `dimensionFilters` and `dimensions` arguments should not have any relation. `dimensionFilters` are a standalone argument.

**Review by felixarntz — CHANGES_REQUESTED** [review]
I believe there's something off in the implementation here. Why is `dimensionFilters` validation/normalization somehow based on the passed `dimensions`? These are separate arguments with entirely different purposes. Looks like I went over the IB too quickly 🤦‍♂️ 

cc @aaemnnosttv @eugene-manuilov

**Comment by ivankruchkoff** [comment]
Updating, per your comments @felixarntz. The IB had:
```
 It should check whether dimensionFilters is a map of dimension names and values and all dimension names are also defined in the dimensions option.
```

File: includes/Modules/Analytics.php (hunk 19 above)
Comment (ivankruchkoff) [reply includes/Modules/Analytics.php]: Updated.

File: includes/Modules/Analytics.php (hunk 20 above)
Comment (ivankruchkoff) [reply includes/Modules/Analytics.php]: Updated.

File: includes/Modules/Analytics.php (hunk 21 above)
Comment (ivankruchkoff) [reply includes/Modules/Analytics.php]: Updated.

File: assets/js/modules/analytics/util/report-validation.js (hunk 22 above)
Comment (ivankruchkoff) [reply assets/js/modules/analytics/util/report-validation.js]: Updated.

File: assets/js/modules/analytics/util/report-normalization.js [hunk 23]
```diff
@@ -51,3 +53,26 @@ const normalizeMetrics = ( metrics ) => {
...
+const normalizeDimensions = ( dimensions ) => {
+	return castArray( dimensions )
+		.map(
+			( dimension ) => typeof dimension === 'string'
+				? { name: dimension }
+				: dimension
+		)
+		.filter( ( dimension ) => isPlainObject( dimension ) )
+	;
+};
```
Comment (felixarntz) [inline assets/js/modules/analytics/util/report-normalization.js:66]: The normalization here makes sense, but this causes a bug visible on the dashboard, in `DashboardPopularPagesWidget`, where (for whatever odd reason 😆 ) we're passing multiple values as a comma-separated string. I'll update that component to pass the values as array of strings which is more appropriate.

File: assets/js/modules/analytics/util/report-normalization.js [hunk 24]
```diff
@@ -51,3 +53,26 @@ const normalizeMetrics = ( metrics ) => {
...
+	const normalizedFilters = {};
+	if ( isPlainObject( dimensionFilters ) ) {
+		for ( const [ key, value ] of Object.entries( dimensionFilters ) ) {
+			if ( typeof value === 'string' ) {
+				normalizedFilters[ key ] = value;
+			}
+		}
+	}
+	return normalizedFilters;
+};
```
Comment (felixarntz) [inline assets/js/modules/analytics/util/report-normalization.js:78]: I don't think we need a normalization function for dimension filters. The reason they exist for dimensions and metrics is that those arguments support multiple different ways of being specified (e.g. string, array of strings, array of objects). For dimension filters though, for now we only support a map of scalar values - so let's remove `normalizeDimensionFilters`. We only need validation here.

File: assets/js/modules/analytics/util/report-validation.js [hunk 25]
```diff
@@ -57,3 +57,18 @@ export function isValidDimensions( dimensions ) {
...
+ * @since n.e.x.t
+ *
+ * @param {Object} dimensionFilters The dimension filters to check.
+ * @return {boolean} TRUE if dimension filters are valid, otherwise FALSE.
+ */
+export function isValidDimensionFilters( dimensionFilters ) {
+	// Ensure every dimensionFilter key corresponds to a valid dimension.
+	return Object.keys( dimensionFilters ).every(
+		( dimension ) => typeof dimensionFilters[ dimension ] === 'string'
+	);
```
Comment (felixarntz) [inline assets/js/modules/analytics/util/report-validation.js:73]: We shouldn't require dimension values to be a string. They just need to be scalar values (probably we could check whether it's either a string or a number). Something more important to cover in the validation here though is to ensure that a map of `dimensionName => dimensionValue` is passed. The keys here actually need to be strings. For example someone shouldn't be able to pass an array of values.

File: assets/js/modules/analytics/datastore/report.js [hunk 26]
```diff
@@ -57,20 +57,23 @@ const fetchGetReportStore = createFetchStore( {
...
-		if ( dimensions ) {
-			invariant(
-				isValidDimensions( dimensions ),
-				'Dimensions for an Analytics report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "name" property.',
-			);
-		}
+		invariant(
+			isValidDimensions( dimensions ),
+			'Dimensions for an Analytics report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "name" property.',
+		);
```
Comment (felixarntz) [inline assets/js/modules/analytics/datastore/report.js:71]: Why did you remove the if clause here? I think that's an unrelated change.

File: assets/js/modules/analytics/datastore/report.js [hunk 27]
```diff
@@ -57,20 +57,23 @@ const fetchGetReportStore = createFetchStore( {
...
+
+		invariant(
+			isValidDimensionFilters( dimensionFilters ),
+			'Dimension filters must be an object where the values are strings.',
```
Comment (felixarntz) [inline assets/js/modules/analytics/datastore/report.js:75]: See above comment, let's update the messaging here to be more appropriate, e.g.
```suggestion
			'Dimension filters must be a map of dimension names as keys and dimension values as values.',
```

**Review by felixarntz — CHANGES_REQUESTED** [review]
Mostly LGTM, a few minor follow-up things to address.

File: assets/js/modules/analytics/datastore/report.js (hunk 26 above)
Comment (ivankruchkoff) [reply assets/js/modules/analytics/datastore/report.js:71]: Yes, unrelated, thanks for noticing it.

File: assets/js/modules/analytics/util/report-normalization.js (hunk 23 above)
Comment (ivankruchkoff) [reply assets/js/modules/analytics/util/report-normalization.js:66]: We initially had the normalisation to pass dimensions to the `isValidDimensionFilters`, which we don't need now, but it's a good idea to keep it as it makes it cleaner to have it normalised.

File: assets/js/modules/analytics/util/report-validation.js (hunk 25 above)
Comment (ivankruchkoff) [reply assets/js/modules/analytics/util/report-validation.js:73]: I figured dimension values would be string due to https://developers.google.com/analytics/devguides/reporting/core/v4/rest/v4/reports/batchGet#DimensionFilter ( expressions[] are 'string' )

Comment (ivankruchkoff) [reply assets/js/modules/analytics/util/report-validation.js:73]: I've added a check for both string and number per your comment, and checking that keys are string.

**Review by felixarntz — APPROVED** [review]
Awesome!
